src/overlays/effects/ovl_Effect_Ss_Hahen/z_eff_ss_hahen.c

```diff
@@ -5,29 +5,127 @@
  */
 
 #include "z_eff_ss_hahen.h"
+#include "objects/gameplay_keep/gameplay_keep.h"
+
+#define rPitch regs[0]
+#define rYaw regs[1]
+#define rFlags regs[2]
+#define rScale regs[3]
+#define rObjId regs[4]
+#define rObjBankIndex regs[5]
+#define rMinLife regs[6]
 
 #define PARAMS ((EffectSsHahenInitParams*)initParamsx)
 
-s32 EffectSsHahen_Init(PlayState* play, u32 index, EffectSs* this, void* initParamsx);
+u32 EffectSsHahen_Init(PlayState* play, u32 index, EffectSs* this, void* initParamsx);
 void EffectSsHahen_Update(PlayState* play, u32 index, EffectSs* this);
 void EffectSsHahen_Draw(PlayState* play, u32 index, EffectSs* this);
 
-#if 0
 const EffectSsInit Effect_Ss_Hahen_InitVars = {
     EFFECT_SS_HAHEN,
     EffectSsHahen_Init,
 };
 
-#endif
+void EffectSsHahen_CheckForObject(EffectSs* this, PlayState* play) {
+    if (((this->rObjBankIndex = Object_GetIndex(&play->objectCtx, this->rObjId)) < 0) ||
+        !Object_IsLoaded(&play->objectCtx, this->rObjBankIndex)) {
+        this->life = -1;
+        this->draw = NULL;
+    }
+}
+
+u32 EffectSsHahen_Init(PlayState* play, u32 index, EffectSs* this, void* initParamsx) {
+    EffectSsHahenInitParams* initParams = PARAMS;
+
+    this->pos = initParams->pos;
+    this->velocity = initParams->velocity;
+    this->accel = initParams->accel;
+    this->life = 200;
+
+    if (initParams->dList != NULL) {
+        this->gfx = initParams->dList;
+        this->rObjId = initParams->objId;
+        EffectSsHahen_CheckForObject(this, play);
+    } else {
+        this->gfx = gEffFragmentsDL;
+        this->rObjId = HAHEN_OBJECT_DEFAULT;
+    }
+
+    this->draw = EffectSsHahen_Draw;
+    this->update = EffectSsHahen_Update;
+    this->rFlags = initParams->flags;
+    this->rScale = initParams->scale;
+    this->rPitch = Rand_ZeroOne() * 314.0f;
+    this->rYaw = Rand_ZeroOne() * 314.0f;
+    this->rMinLife = 200 - initParams->life;
+
+    return 1;
+}
+
+void EffectSsHahen_DrawOpa(PlayState* play, EffectSs* this) {
+    GraphicsContext* gfxCtx = play->state.gfxCtx;
+    s32 pad;
+
+    OPEN_DISPS(gfxCtx);
+
+    if (this->rObjId != HAHEN_OBJECT_DEFAULT) {
+        gSPSegment(POLY_OPA_DISP++, 0x06, play->objectCtx.status[this->rObjBankIndex].segment);
+    }
+    gSPMatrix(POLY_OPA_DISP++, Matrix_NewMtx(gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    func_8012C28C(play->state.gfxCtx);
+    gSPDisplayList(POLY_OPA_DISP++, this->gfx);
+
+    CLOSE_DISPS(gfxCtx);
+}
+
+void EffectSsHahen_DrawXlu(PlayState* play, EffectSs* this) {
+    GraphicsContext* gfxCtx = play->state.gfxCtx;
+    s32 pad;
```

Comment:
Don't suppose there's any indiction from elsewhere of what this could be?

Reply:
It is possibly a PlayState upcast, since draw doesn't necessarily upcast, but could have just passed GameState straight on.